src/hotspot/share/opto/loopTransform.cpp

```diff
@@ -1595,6 +1683,166 @@ void PhaseIdealLoop::insert_vector_post_loop(IdealLoopTree *loop, Node_List &old
   // finds some, but we _know_ they are all useless.
   peeled_dom_test_elim(loop, old_new);
   loop->record_for_igvn();
+  C->print_method(PHASE_AFTER_ATOMIC_POST, 4, post_head);
+}
+
+//------------------------------insert_atomic_post_loop_impl-------------------------------
+// The main implementation of inserting atomic post loop after vector main loop.
```

Comment:
I would really appreciate some kind of ascii art here. It should show the pre, main, vectorized-post and post loop.
And the relevant zero-trip guards, etc.

Reply:
And maybe you can also draw how the rewiring happens, i.e. where you cut/glue the graph back together.

Reply:
I think that will make reviewing much easier :)

Like maybe trace the trip phi, and some other imaginary/memory data phi.
The rewiring here looks tricky, and I'll have to spend some time looking at it.